File: advent_of_code/2024/python/04.py
directions = [
    (dx, dy) for dx in range(-1, 2) for dy in range(-1, 2) if dx != 0 or dy != 0
]


def get_word_at(position, direction, size, grid):
    x, y = position
    dx, dy = direction
    return ''.join(
        [
            grid[x + dx * i][y + dy * i]
            for i in range(size)
            if 0 <= x + dx * i < len(grid) and 0 <= y + dy * i < len(grid[0])
        ]
    )


def part1(text):
    grid = text.splitlines()
    n, m = len(grid), len(grid[0])
    xmas_count = 0
    for x in range(n):
        for y in range(m):
            for d in directions:
                xmas_count += get_word_at((x, y), d, 4, grid) == 'XMAS'
    return xmas_count


def part2(text):
    grid = text.splitlines()
    n, m = len(grid), len(grid[0])
    xmas_count = 0
    for x in range(n):
        for y in range(m):
            if x + 2 < n and y + 2 < m:
                xmas_count += get_word_at((x, y), (1, 1), 3, grid) in [
                    'MAS',
                    'SAM',
                ] and get_word_at((x + 2, y), (-1, 1), 3, grid) in [
                    'MAS',
                    'SAM',
                ]
    return xmas_count

File: advent_of_code/2024/rust/src/04.rs
use std::collections::HashSet;

fn directions() -> Vec<(isize, isize)> {
    vec![
        (-1, -1),
        (-1, 0),
        (-1, 1),
        (0, -1),
        (0, 1),
        (1, -1),
        (1, 0),
        (1, 1),
    ]
}

fn get_word_at(
    position: (usize, usize),
    direction: (isize, isize),
    size: usize,
    grid: &[Vec<char>],
) -> Option<String> {
    let (x, y) = (position.0 as isize, position.1 as isize);
    let (dx, dy) = direction;
    let mut word = String::new();

    for i in 0..size as isize {
        let nx = x + dx * i;
        let ny = y + dy * i;
        if nx < 0 || ny < 0 || nx >= grid.len() as isize || ny >= grid[0].len() as isize {
            return None;
        }
        word.push(grid[nx as usize][ny as usize]);
    }
    Some(word)
}

fn part1(text: &str) -> usize {
    let grid: Vec<Vec<char>> = text.lines().map(|line| line.chars().collect()).collect();
    let directions = directions();
    let mut xmas_count = 0;

    for x in 0..grid.len() {
        for y in 0..grid[0].len() {
            for &d in &directions {
                if let Some(word) = get_word_at((x, y), d, 4, &grid) {
                    if word == "XMAS" {
                        xmas_count += 1;
                    }
                }
            }
        }
    }

    xmas_count
}

fn part2(text: &str) -> usize {
    let grid: Vec<Vec<char>> = text.lines().map(|line| line.chars().collect()).collect();
    let mas_variants: HashSet<&str> = ["MAS", "SAM"].iter().cloned().collect();
    let mut xmas_count = 0;

    for x in 0..grid.len() {
        for y in 0..grid[0].len() {
            if x + 2 < grid.len() && y + 2 < grid[0].len() {
                let start_pos = (x, y);
                let check_pos = (x + 2, y);
                if let (Some(diag1), Some(diag2)) = (
                    get_word_at(start_pos, (1, 1), 3, &grid),
                    get_word_at(check_pos, (-1, 1), 3, &grid),
                ) {
                    if mas_variants.contains(diag1.as_str())
                        && mas_variants.contains(diag2.as_str())
                    {
                        xmas_count += 1;
                    }
                }
            }
        }
    }

    xmas_count
}

fn main() {
    aoc2024::solve(part1, part2);
}
